storage/engine,libroach: 1x write amplification on ingest sst

dt opened the pull request:
This uses the new flag added in https://github.com/facebook/rocksdb/pull/4172 to avoid needing to make global_seq_no-related edits to SSTs, and thus avoid the SST copying those edits forced us to do to preserve the raft log immutability when hard-linking directly to side load SSTs.

This is gated on a cluster version that is defined well after the upgrade to rocks 5.17, since using the flag requires that no older versions of RocksDB (<5.16),
will ever be used to read these SSTs, as they will lack the seq_no that the older versions need for correctness.

Release note (performance improvement): reduce write-amplification during bulk-loading (IMPORT and RESTORE)

dt committed:
storage/engine,libroach: 1x write amplification on ingest sst

This uses he new flag added in
https://github.com/facebook/rocksdb/pull/4172 to avoid the global_seq_no
edits and the SST copying they forced us to do to preserve the raft log.

Passing this flag has to be gated on a cluster version that is defined
well after the upgrade to rocks 5.17 -- using the flag requires that
older versions of RocksDB (< 5.16) will never be used to read, as they
will still be looking for and using the seq_no in the file.

Release note (performance improvement): reduce write-amplification during bulk-loading (IMPORT and RESTORE)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34886)
<!-- Reviewable:end -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Were you able to measure a perf benefit or otherwise verify that this reduces disk traffic during IMPORT? Or perhaps I'd be satisfied if we could have a unit test that this is working properly. I'm not quite sure how that would work, but perhaps generate an SST and ingest it and verify that the SST is unmodified. Somehow. Maybe.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34886#-:-LYgTsfJ9hxlJDe57E2Q:b-c9v6av)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @petermattis, and @tbg)

___
*[c-deps/libroach/db.cc, line 700 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34886#-LYgTglU3tYQfth_JeUV:-LYgTglVBAX1f19TKRPL:b-bbl5fb) ([raw file](https://github.com/cockroachdb/cockroach/blob/56a47a1b7fb3c98e8a855e55dfc67975d4091b06/c-deps/libroach/db.cc#L700)):*
> ```cpp
>   // only safe to disable the seqno if older versions of RocksDB (<5.16) will
>   // not be used to read these SSTs.
>   // Before editing a file to insert a seq_no, RocksDB also checks the option
> ```

Nit: add a blank line between paragraphs.


<!-- Sent from Reviewable.io -->

dt commented:
Realized it wasn't working because rocks checks the allow_modification flag before it checks if it is actually using seqno at all. Added a `|| skip_seqno` to the allow_modification flag indeed, it seems to be working as expected. On a very fast MBP SSD that is otherwise unloaded, the speedup on a direct (overlapping) import of ~1.7gb of tpcc data looks like about 5%, but past experience is that using available disk IO is a bigger deal on more throughput constrained disks (e.g. ebs) or with other traffic contending for disk io. I observed the process wrote about 1.5gb less, but with overlapping import, the heavy recompactions dominate total write traffic so it only worked out to about a 20% reduction in total write traffic.

dt commented:
@petermattis turns out we _did_ actually have a unit test this that should have started failing if the original change had worked as intended -- I found (and updated it) when I fixed it. Here's the sst apply vs copy with and without this change enabled:
<img width="526" alt="screen shot 2019-02-21 at 2 37 12 pm" src="https://user-images.githubusercontent.com/15615/53198334-2b80ee00-35ea-11e9-87db-394e64a6d558.png">

petermattis reviewed (commented):
Looks good. Minor nit remaining.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34886#-:-LZI1ld6AmEf8UURPfI4:by2x3e)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt and @tbg)

___
*[c-deps/libroach/db.cc, line 700 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34886#-LYgTglU3tYQfth_JeUV:-LZI1B_r5gFUGQZyDaok:b-82oqa1) ([raw file](https://github.com/cockroachdb/cockroach/blob/56a47a1b7fb3c98e8a855e55dfc67975d4091b06/c-deps/libroach/db.cc#L700)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit: add a blank line between paragraphs.
</blockquote></details>

Ping.
___
*[c-deps/libroach/include/libroach.h, line 412 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34886#-LZI1S_72T1gtQYyTsQu:-LZI1S_72T1gtQYyTsQv:b-n05qje) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7c78bb20c25505636f80d93859b4893041c8f71/c-deps/libroach/include/libroach.h#L412)):*
> ```objc
> // it would have tried to modify any of the files' sequence numbers rather than
> // editing the files in place. If skip_seqno is true, it will skip the
> // seq_no entirely -- this is only safe if this will only ever be read by Rocks
> ```

This is a bit confusing. `skip_seqno` indicates that the global seqno doesn't need to be written into the sstable. The global seqno is written into the LSM state, just outside of the sstable. I'd prefer to keep this parameter named as `write_global_seqno`, and to clarify the commentary.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34886)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @petermattis and @tbg)

___
*[c-deps/libroach/db.cc, line 700 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34886#-LYgTglU3tYQfth_JeUV:-LZLmG8H4rBkvbC4MLU2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56a47a1b7fb3c98e8a855e55dfc67975d4091b06/c-deps/libroach/db.cc#L700)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ping.
</blockquote></details>

Done.
___
*[c-deps/libroach/include/libroach.h, line 412 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34886#-LZI1S_72T1gtQYyTsQu:-LZLmFD_3MSX3emuFIaN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7c78bb20c25505636f80d93859b4893041c8f71/c-deps/libroach/include/libroach.h#L412)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is a bit confusing. `skip_seqno` indicates that the global seqno doesn't need to be written into the sstable. The global seqno is written into the LSM state, just outside of the sstable. I'd prefer to keep this parameter named as `write_global_seqno`, and to clarify the commentary.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dt commented:
bors r+
